refactor: 1. introduce dataEngine [TECH-344]

JoakimSM opened the pull request:
First part of [TECH-344](https://jira.dhis2.org/browse/TECH-344)

This PR makes the query function from the data engine in the app-runtime available to our metadata loader. I wanted to avoid using a global for this, later when we're going to use it in the epics we can DI it directly.

Some of the code earlier found in entry.js is moved to a new React compoent file AppLoader.component to be able to use the hook useDataEngine from the app-platform.

Most of the changes are just moving code around, trying to stay on point and not fixing everything else (I did fix some linting and flow errors).

paschalidi commented on src/init/entry.js:
```diff
@@ -3,31 +3,22 @@
 import { render } from 'react-dom';
 import log from 'loglevel';
-import i18n from '@dhis2/d2-i18n';
 import 'typeface-roboto';
```
since we are touching this file I have a suggestion. 

Could we move this line and the line about `import './addRxjsOperators';` to [here](https://github.com/dhis2/capture-app/blob/refactor-metadata-fetching/src/index.js)?

Looks like this is a more applicable place for these kind of imports

JoakimSM replied:
Pretty sure at one point it was imported in the file you're mentioning. So I'm a bit wary about moving it without properly testing it.

paschalidi replied:
Okay good point. Thats out of scope as well. Maybe I could open a PR at some point with all those suggestions.

paschalidi commented on src/init/entry.js:
```diff
@@ -73,50 +59,40 @@ function handleCacheExpired(domElement: HTMLElement) {
-} else {
-    loadAppAsync(domElement);
-}
+loadAppAsync();
```
Why we dont export this function so that we can call it on this [file](https://github.com/dhis2/capture-app/blob/refactor-metadata-fetching/src/index.js)?

Every person who would start reading this code from scratch and knows react would fo into that file. When you first  go there as completely new to the project, as it is now, you see just some imports and you have to look into all those files to find where the app is actually starting from. I would argue that will help people who are gonna join us in the future to better understand what is happening here. 

What is your stand here Joakim?

JoakimSM replied:
Making an exception to my stay on point rule for this suggestion. Thanks 👍

JoakimSM commented on i18n/en.pot:
```diff
@@ -5,8 +5,8 @@ msgstr ""
 "Plural-Forms: nplurals=2; plural=(n != 1)\n"
-"POT-Creation-Date: 2020-03-09T19:35:56.143Z\n"
-"PO-Revision-Date: 2020-03-09T19:35:56.143Z\n"
+"POT-Creation-Date: 2020-05-08T10:24:17.432Z\n"
```
This is an autogenerated file for i18n, created when you run/build the app. I tried to gitignore it at some point but ran into trouble so let's just live with it for the time being.

paschalidi replied:
Arent those generated and merged to master automatically? In case this is true. 

I would discard the changes that has been introduced here and I would expect to `master` to  have those changes already when we merge back to it.
Am I missing something?

JoakimSM replied:
The pot file is not merged to master automatically, I believe your thinking about the po files.

paschalidi replied:
yes you right, I dont really know how these things work tbh. maybe I need to ask someone to give me an intro at some point.

JoakimSM commented on src/components/App/App.component.js:
```diff
@@ -1,5 +1,5 @@
 // @flow
```
I agree about the EventCaptureForm, but I would like to keep the components folder. 1. Because it separates the React components from the other stuff 2. The capture-core is also structured this way. 

Let's leave it for another issue though (trying to stay on point with the issue at hand here).

JoakimSM commented on src/components/App/App.component.js:
```diff
@@ -1,5 +1,5 @@
 // @flow
-/* eslint-disable */
+/* eslint-disable import/first */
```
The reason I didn't: this would change the insertion order of the css into the DOM. Right now the css from app.css is inserted before everything else in the file and the files dependencies. It think it makes sense to insert the app.css file first in case you want to override something.

JoakimSM commented on src/core_modules/capture-core/components/LoadingMasks/index.js:
```diff
@@ -0,0 +1,3 @@
+// @flow
+export { default as LoadingMask } from './LoadingMask.component';
+export { default as LoadingMaskForPage } from './LoadingMaskForPage.component';
```
I agree, but didn't want to do it in this PR. Let's leave the these consistency remarks unresolved and I will create a new PR for those.

JoakimSM commented on src/core_modules/capture-core/metaDataStoreLoaders/index.js:
```diff
@@ -1,3 +1,3 @@
 // @flow
-export { default as loadMetaDataAsync } from './baseLoader/metaDataLoader';
-export { default as loadSystemSettingsAsync } from './systemSettings/loadSystemSetttings';
+export { default as loadMetaData } from './baseLoader/metaDataLoader';
```
I agere with most of your points here. Let's see:
1. > The first is about the default as loadMetaData. Can we say that we will have one name per function even if we do stuff like default as functionName? For example here I would rename the loadMetaDataAsync to loadMetaData inside the metaDataLoader.js file. This way the function has one name always. Is this sensible to you?

For the exact function mentioned here, loadMetaDataAsync is being renamed to loadMetData in a later PR in this PR-chain, so I agree with your point in principle. Sometimes the function needs renaming in the export due to the context though, but this is not the case here.

**While we are on the subject: I've stoppped exporting as default altogether. This makes refactoring easier since you have predictable names in the import. So I would suggest we keep this practice. Up for discussion if you disagree.**   

2. > The second has to do with the name of the folder metaDataLoader.js. Can we for simplicity name this folder loadMetaData.js as we do bellow with the loadSystemSetttings in the line bellow?

I guess you're talking about the file name here (not folder name)? If that is the case, I agree.

paschalidi replied:
> While we are on the subject: I've stoppped exporting as default altogether. This makes refactoring easier since you have predictable names in the import. So I would suggest we keep this practice. Up for discussion if you disagree.

Yes agree. 

> I guess you're talking about the file name here (not folder name)? If that is the case, I agree.

Yes I was talking about the file name :)

JoakimSM commented on src/core_modules/capture-core/metaDataStoreLoaders/index.js:
```diff
@@ -1,3 +1,3 @@
-export { default as loadMetaDataAsync } from './baseLoader/metaDataLoader';
-export { default as loadSystemSettingsAsync } from './systemSettings/loadSystemSetttings';
+export { default as loadMetaData } from './baseLoader/metaDataLoader';
+export { default as loadSystemSettings } from './systemSettings/loadSystemSetttings';
```
I agree (I regret touching the export in this PR, but cause I've already started it, I'll fix it  👍)

paschalidi replied:
Thanks a lot!

JoakimSM commented on src/core_modules/capture-core/utils/exceptions/index.js:
```diff
@@ -0,0 +1,2 @@
+// @flow
+export { DisplayException } from './DisplayException';
```
Yes 👍 Please keep this unresolved, I'll do it in the consistency PR later.

paschalidi replied:
Good idea. Will be keeping it unresolved

JoakimSM commented on src/init/AppLoader.component.js:
```diff
@@ -0,0 +1,72 @@
+// @flow
```
Hmm, maybe not the App folder, but I agree it should have been restructured. Let me think about it and maybe do something in the last PR for the issue.

paschalidi replied:
okay good stuff. Lets follow this up on the sixth PR! 
I will keep it unresolved for now.

paschalidi replied:
Thanks for looking into it!

JoakimSM commented on src/init/AppLoader.component.js:
```diff
@@ -0,0 +1,72 @@
+
+    const load = React.useCallback(async () => {
+        try {
+            await initializeAsync({
```
Meaning: I will make the function take two arguments 🙂

paschalidi replied:
Great thanks!

paschalidi replied:
Have you gotten the chance to look into it?

JoakimSM commented on src/init/AppLoader.component.js:
```diff
@@ -0,0 +1,72 @@
+                    message += ' Please see log for details.';
+                }
+            }
+            setLoadError(message);
```
I like it and appreciate the suggestion 👍 But to keep this PR focused on the issue at hand I will try to be selective about what I'm fixing (this is "copy paste" code from another file).

paschalidi replied:
agree.

JoakimSM commented on src/init/entry.js:
```diff
@@ -37,6 +28,15 @@ const generateClassName = createGenerateClassName();
+// DOM element to use as the container for the App
+const DOM_ID = 'app';
+const domElement = document.getElementById(DOM_ID);
+if (!domElement) {
```
You're right 🙂

paschalidi replied:
I know you moved the code around so this is not the scope of this PR. I am resolving this then.

JoakimSM commented:
I think I've addressed your comments now. Let me know if I'm missing some.

>One question @JoakimSM, this PR isnt pointing onto master but to another branch. Since this is the first PR of the stacked PRs should it be pointing to master? Am I missing something?

Answer from slack:  I did it this way so we wouldn’t accidentally merge the branch into master before I was done. Will create a PR for master from that branch when everything is settled.
